core/registry.js

```diff
@@ -0,0 +1,174 @@
+      ' non-empty string.');
+  }
+
+  if ((typeof name != 'string') || (name.trim() == '')) {
+    throw Error('Invalid name "' + name + '". The name must be a' +
+      ' non-empty string.');
+  }
+  if (!registryClass) {
+    throw Error('Can not register a null value');
+  }
+  var typeRegistry = Blockly.registry.typeMap_[type.toLowerCase()];
+  // If the type registry has not been created, create it.
+  if (!typeRegistry) {
+    typeRegistry = Blockly.registry.typeMap_[type.toLowerCase()] = {};
+  }
+
+  // Validate that the given class has all the required properties.
+  Blockly.registry.validate_(type, registryClass);
+
+  // If the name already exists throw an error.
+  if (typeRegistry[name.toLowerCase()]) {
+    throw Error('Name "' + name + '" with type "' + type + '" already registered.');
+  }
+  typeRegistry[name.toLowerCase()] = registryClass;
+};
+
+/**
+ * Checks the given class for properties that are required based on the type.
+ * @param {string} type The type of the plugin. (eg: Field, Renderer)
+ * @param {Function} registryClass A class that we are checking for the required
+ *    properties.
+ * @private
+ */
+Blockly.registry.validate_ = function(type, registryClass) {
+  switch (type) {
+    case String(Blockly.registry.Type.FIELD):
+      Blockly.registry.validateProperties_(type, ['fromJson'], registryClass);
+      break;
+    case String(Blockly.registry.Type.RENDERER):
+      // TODO: Figure out what methods to check here.
+      Blockly.registry.validateProperties_(type, [], registryClass);
+      break;
+  }
+};
+
+/**
+ * Checks that the given class has all the given method names.
+ * @param {string} type The type of the plugin. (eg: Field, Renderer)
+ * @param {Array.<string>} requiredProperties The list of method names we expect the
+ *     given class to have.
+ * @param {Function} registryClass A class that we are checking for the required
+ *     properties.
+ * @throws {Error} if the class does not implement all of the method names.
+ * @private
+ */
+Blockly.registry.validateProperties_ = function(type, requiredProperties, registryClass) {
+  var unimplemented = requiredProperties.filter(function(property) {
+    return !registryClass.hasOwnProperty(property) ||
+        (typeof registryClass[property] != 'function');
+  });
+  if (unimplemented.length) {
+    throw Error('Type "' + type + '" requires the following properties "' +
+        unimplemented + '"');
+  }
+};
+
+/**
+ * Unregisters the class with the given type and name.
+ * @param {string|Blockly.registry.Type<T>} type The type of the plugin.
+ *     (eg: Field, Renderer)
+ * @param {string} name The plugin's name. (Ex. field_angle, geras)
+ * @template T
+ */
+Blockly.registry.unregister = function(type, name) {
+  type = String(type);
+  var typeRegistry = Blockly.registry.typeMap_[type.toLowerCase()];
+  if (!typeRegistry) {
+    console.warn('No type "' + type + '" found');
+    return;
+  }
+  if (!typeRegistry[name]) {
```

Comment:
Name is not normalized before this check, why not just:
```
type = String(type).toLowerCase()
name = name.toLowerCase()
```
at the top? If it's so the error message reads the value that the developer gave, I would suggest otherwise (better that the developer see what value we're checking against).